### PR TITLE
[Merged by Bors] - feat(CategoryTheory): Beck's comonadicity theorem

### PR DESCRIPTION
**Pull request by mckoen**
Prove Beck's comonadicity theorem by dualizing everything in [CategoryTheory.Monad.Monadicity](https://leanprover-community.github.io/mathlib4_docs/Mathlib/CategoryTheory/Monad/Monadicity.html).

This contribution was created as part of the AIM workshop "Formalizing Algebraic Geometry" in June 2024.

---

- [x] depends on: #14170 
- [x] depends on: #14171 
- [x] depends on: #14236 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Monad/Comonadicity.lean [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+  apply IsCoreflexivePair.mk' (G.map (adj.counit.app _)) _ _
+  · rw [← G.map_comp, ← G.map_id]
+    exact congr_arg G.map A.counit
+  · rw [adj.right_triangle_components]
+    rfl
+
+/-- The "main pair" for a coalgebra `(A, α)` is the pair of morphisms `(F α, ε_FA)`. It is always a
+`G`-cosplit pair, and will be used to construct the right adjoint to the comparison functor and show
+it is an equivalence.
+-/
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Monad/Comonadicity.lean:69]:
```suggestion
/-- The "main pair" for a coalgebra `(A, α)` is the pair of morphisms `(G α, η_GA)`. It is always a
`G`-cosplit pair, and will be used to construct the right adjoint to the comparison functor and show
it is an equivalence.
-/
```

File: Mathlib/CategoryTheory/Monad/Comonadicity.lean [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+    dsimp
+    infer_instance
+  · apply @preservesLimitOfIsoDiagram _ _ _ _ _ _ _ _ _ (diagramIsoParallelPair.{v₁} _).symm ?_
+    dsimp
+    infer_instance
+
+section BeckComonadicity
+
+-- [∀ ⦃A B⦄ (f g : A ⟶ B) [F.IsCosplitPair f g], HasEqualizer f g]
+/-- Dual to `Monad.HasCoequalizerOfIsSplitPair`. -/
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Monad/Comonadicity.lean:233]:
```suggestion
/-- Dual to `Monad.HasCoequalizerOfIsSplitPair`. -/
```

File: Mathlib/CategoryTheory/Monad/Comonadicity.lean [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+  /-- If `f, g` is an `F`-cosplit pair, then they have an equalizer. -/
+  out : ∀ {A B} (f g : A ⟶ B) [F.IsCosplitPair f g], HasEqualizer f g
+
+instance [HasEqualizerOfIsCosplitPair F] : ∀ (A : Coalgebra adj.toComonad),
+    HasEqualizer (G.map A.a)
+      (adj.unit.app (G.obj A.A)) :=
+  fun _ => HasEqualizerOfIsCosplitPair.out F _ _
+
+-- [∀ ⦃A B⦄ (f g : A ⟶ B) [F.IsCosplitPair f g], PreservesLimit (parallelPair f g) F]
+/-- Dual to `Monad.PreservesColimitOfIsSplitPair`. -/
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Monad/Comonadicity.lean:244]:
```suggestion
/-- Dual to `Monad.PreservesColimitOfIsSplitPair`. -/
```

File: Mathlib/CategoryTheory/Monad/Comonadicity.lean [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+instance {A B} (f g : A ⟶ B) [F.IsCosplitPair f g] [PreservesLimitOfIsCosplitPair F] :
+    PreservesLimit (parallelPair f g) F := PreservesLimitOfIsCosplitPair.out f g
+
+instance [PreservesLimitOfIsCosplitPair F] : ∀ (A : Coalgebra adj.toComonad),
+   PreservesLimit (parallelPair (G.map A.a)
+      (NatTrans.app adj.unit (G.obj A.A))) F :=
+  fun _ => PreservesLimitOfIsCosplitPair.out _ _
+
+-- [∀ ⦃A B⦄ (f g : A ⟶ B) [F.IsCosplitPair f g], ReflectsLimit (parallelPair f g) F] :
+/-- Dual to `Monad.ReflectsColimitOfIsSplitPair`. -/
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Monad/Comonadicity.lean:258]:
```suggestion
/-- Dual to `Monad.ReflectsColimitOfIsSplitPair`. -/
```

File: Mathlib/CategoryTheory/Monad/Comonadicity.lean [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+      apply @unitEqualizerOfCoreflectsEqualizer _ _ _ _ _ _ _ _ ?_
+      letI _ :
+        F.IsCosplitPair (G.map (F.map (adj.unit.app Y)))
+          (adj.unit.app (G.obj (F.obj Y))) :=
+        ComonadicityInternal.main_pair_F_cosplit _ ((comparison adj).obj Y)
+      infer_instance
+    exact (comparisonAdjunction adj).toEquivalence.symm.isEquivalence_inverse
+
+-- [∀ ⦃A B⦄ (f g : A ⟶ B) [F.IsCosplitPair f g], CreatesLimit (parallelPair f g) F] :
+/-- Dual to `Monad.CreatesColimitOfIsSplitPair`. -/
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Monad/Comonadicity.lean:303]:
```suggestion
/-- Dual to `Monad.CreatesColimitOfIsSplitPair`. -/
```

File: Mathlib/CategoryTheory/Monad/Comonadicity.lean [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+  apply comonadicOfHasPreservesReflectsFSplitEqualizers adj
+
+end BeckComonadicity
+
+section CoreflexiveComonadicity
+
+variable [HasCoreflexiveEqualizers C] [F.ReflectsIsomorphisms]
+
+-- [∀ ⦃A B⦄ (f g : A ⟶ B) [F.IsCoreflexivePair f g], PreservesLimit (parallelPair f g) F] :
+/-- Dual to `Monad.PreservesColimitOfIsReflexivePair`. -/
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Monad/Comonadicity.lean:349]:
```suggestion
/-- Dual to `Monad.PreservesColimitOfIsReflexivePair`. -/
```

**Review by dagurtomas — APPROVED** [review]
Please add comments in the module docstrings of both files `Comonadicity` and `Monadicity` to keep them in sync. 

You should also merge master and make sure everything works with the new variable incusion mechanism. Otherwise LGTM.

**Comment by dagurtomas** [comment]
Thanks!

maintainer merge

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**